### ISSUE TITLE
Provide an option to disable completion in comments

### ISSUE THREAD

**Issue by ville-h** [racer-rust/emacs-racer#6]
Emacs stalls while writing in some comments such as the comment: "// writing in this comment is broken" below. Try extending the comment with writing words longer than company-minimum-prefix-length and company-idle-delay "low-enough". Simply start tapping the "a" key repeatedly for example and the letters should appear in chunks. Not sure if this is emacs-racer or racer problem and what repo the issue should have been filed to.

``` rust
// company-idle-delay 0.1 is low-enough on my system
fn main() {
    // writing in this comment is fine
}


struct A {
}


impl A {
    fn f() {
        // writing in this comment is broken
        let x = 0;
    }


    fn g() {
        // writing in this comment is fine
    }
}
```

**Comment by ane** [comment]
It seems racer will be called if we're inside a comment. That shouldn't be hard to fix.

**Comment by ville-h** [comment]
If the only solution to fix the stalling is to turn off completons in comments then make it an on-off setting for those who haven't run into this? Some may find completions inside comments useful. I found it rather curious that the slowdown isn't constant in all comments. If you remove the "let x = 0" from above then the "// writing in this comment is broken" no longer stalls emacs.

**Comment by ane** [comment]
Good point. It might be a racer issue that it stalls. Completion is currently disabled only inside strings.

**Comment by Wilfred** [comment]
Setting company-idle-delay to 0.1 is very aggressive. This assumes that all your company back ends together (plus GC, especially if backends are returning many candidates) take less than 0.1 seconds.

We could certainly look at improving performance, but we may not be able to reach 0.1 seconds in all cases.

We could disable completion in comments, but it is a useful feature so it'd have to be an option.

Can you confirm the issue still occurs when racer is the only back end in company-backends? I'm aware of other backends that behave similarly with an aggressive idle delay.

**Comment by ville-h** [comment]
What exactly do you want company-backends to be set to?

**Comment by ane** [comment]
I investigated it a bit further. I made sure `completion-at-point-functions` was only `racer-complete-at-point` and `company-backends` was `company-capf`.

Then, I used the following macro:

```emacs
(defmacro with-timer (title &rest forms)
  "Run the given FORMS, counting the elapsed time.
A message including the given TITLE and the corresponding elapsed
time is displayed."
  (declare (indent 1))
  (let ((nowvar (make-symbol "now"))
        (body   `(progn ,@forms)))
    `(let ((,nowvar (current-time)))
       (message "%s..." ,title)
       (prog1 ,body
         (let ((elapsed
                (float-time (time-subtract (current-time) ,nowvar))))
           (message "%s... done (%.3fs)" ,title elapsed))))))
```
and wrapped it around the call to `->>` in `racer-complete`. Using ville-h's snippet above, I placed the cursor at the start of the comment `// |` and hit `company-complete`. The timer reports in that location about ~0.3 seconds for execution. At the second location, it produces a result in 0.01 seconds! So clearly there's a performance issue here, and I suspect it's something in racer, not Emacs.

**Comment by Wilfred** [comment]
Thanks @ane! If the issue is in Racer, there's not much we can do here. @ville-h would you like to file a bug on Racer itself?

**Comment by ville-h** [comment]
Ok opened issue under racer[0]. A slight annoyance with regards to what values of company-idle-delay are considered "too low". The idle time shouldn't be reset if the key used to invoke racer-complete-at-point is hit too early. Even 0.1s delay is long enough that i often manage to hit tab to try to complete the common part before the 0.1s has passed, thus cancelling the completion and I am left wondering why it doesn't show up. In other words bad usability.

[0] https://github.com/phildawes/racer/issues/413

**Comment by Wilfred** [comment]
Fixed in #60.

**Comment by ane** [comment]
Finally! 👍
